chronos/bipbuffer.nim

```diff
@@ -0,0 +1,167 @@
+
+import results
+
+type
+  BipPos = object
+    start: Natural
+    finish: Natural
+
+  BipBuffer* = object
+    a, b, r: BipPos
+    data: seq[byte]
+
+proc init*(t: typedesc[BipBuffer], size: int): BipBuffer =
+  ## Creates new Bip Buffer with size `size`.
+  BipBuffer(data: newSeq[byte](size))
+
+template len(pos: BipPos): Natural =
+  pos.finish - pos.start
+
+template zero(pos: var BipPos) =
+  pos = BipPos()
+
+func init(t: typedesc[BipPos], start, finish: Natural): BipPos =
+  BipPos(start: start, finish: finish)
+
+func calcReserve(bp: BipBuffer): tuple[space: Natural, start: Natural] =
+  if len(bp.b) > 0:
+    (Natural(bp.a.start - bp.b.finish), bp.b.finish)
+  else:
+    let spaceAfterA = Natural(len(bp.data) - bp.a.finish)
+    if spaceAfterA >= bp.a.start:
+      (spaceAfterA, bp.a.finish)
+    else:
+      (bp.a.start, Natural(0))
+
+func freeSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of free space in buffer `bp`.
+  var sum: Natural
+  if len(bp.b) > 0:
+    sum += (bp.a.start - bp.b.finish)
+  else:
+    sum += bp.a.start
+  sum += Natural(len(bp.data)) - bp.a.finish
+  sum
+
+func availSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of space available for reserve in buffer `bp`.
+  let (res, _) = bp.calcReserve()
+  res
+
+func usedSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of used space in buffer `bp`.
+  len(bp.b) + len(bp.a)
+
+proc reserve*(bp: var BipBuffer, size: Natural): Result[Natural, cstring] =
+  ## Reserve `size` bytes in buffer. Returns number of bytes reserved.
+  doAssert(size <= len(bp.data))
+  let (availableSpace, reserveStart) = bp.calcReserve()
+  if availableSpace == 0:
+    return err("Not enough space available")
+  let reserveLength = min(availableSpace, size)
+  bp.r = BipPos.init(reserveStart, Natural(reserveStart + reserveLength))
+  ok(availableSpace)
+
+proc reserve*(bp: var BipBuffer): Natural =
+  ## Reserve all available free space in buffer. Returns number of bytes
+  ## reserved.
+  ##
+  ## Note, this procedure will raise Defect if there is no free space available.
+  let (availableSpace, reserveStart) = bp.calcReserve()
+  doAssert(availableSpace > 0, "Not enough space available")
+  bp.r = BipPos.init(reserveStart, Natural(reserveStart + availableSpace))
+  availableSpace
+
+proc reserve*(bp: var BipBuffer,
+              pt, st: typedesc): tuple[data: pt, size: st] =
+  ## Reserve all available free space in buffer. Returns current reserved range
+  ## as pointer of type `pt` and size of type `st`.
+  ##
+  ## Note, this procedure will raise Defect if there is no free space available.
+  let (availableSpace, reserveStart) = bp.calcReserve()
+  doAssert(availableSpace > 0, "Not enough space available")
+  bp.r = BipPos.init(reserveStart, Natural(reserveStart + availableSpace))
+  (cast[pt](addr bp.data[bp.r.start]), cast[st](len(bp.r)))
+
+func getReserve*(bp: var BipBuffer,
+                 pt: typedesc, st: typedesc): tuple[data: pointer, size: cint] =
```

Comment:
`cint`??

Reply:
Primitive removed in https://github.com/status-im/nim-chronos/pull/485/commits/e596388d5c2fa222433da97e96d72040473b2bcf